### DISCUSSION TITLE
SSC325 - Imilab EC3 (CMSXJ25A) - IPC025 board

### DISCUSSION THREAD

**Discussion by iamdavidcz** [linux-chenxing/linux-chenxing.org#63] [General]
Hello there,

I just want to let you know that I am working on booting a modified version of the 4.9.84 vendor kernel on **IPC025** board. So far, I was able to retrieve firmware output:

[Boot](https://pastebin.com/raw/kgwcYz3T)
[Boot with microSD card inserted](https://pastebin.com/raw/G6p45UuH)
[Boot with microSD card inserted containing tf_update.img](https://pastebin.com/raw/04zy6biT)
[IPC025B02_4.0.4_0098.zip](https://cdn.cnbj2.fds.api.mi-img.com/chuangmi-cdn/product/ipc025b02/IPC025B02_4.0.4_0098.zip)

**IPL version**: `g2cd6de2`

**U-Boot version**:
```
U-Boot 2015.01 (Sep 14 2020 - 15:25:11), Build: jenkins-ipc025b02_4_0_4_0097-5

Version: I6ge9af17b
```

**U-Boot envs**:
```
SigmaStar # printenv
baudrate=115200
bootargs=console=ttyS0,115200 root=/dev/mtdblock2 rootfstype=squashfs ro init=/linuxrc LX_MEM=0x3fc6000 mma_heap=mma_heap_name0,miu=0,sz=0x1400000
bootcmd=sf probe 0;sf read 0x22000000 ${sf_kernel_start} ${sf_kernel_size};bootm 0x22000000
bootdelay=0
cpu_part_start=14950000
ethact=sstar_emac
ethaddr=00:30:1b:ba:02:db
fileaddr=23b82cf8
filesize=109
gatewayip=172.17.190.1
ipaddr=172.17.190.5
netmask=255.255.255.0
serverip=172.17.190.64
sf_kernel_size=200000
sf_kernel_start=50000
sf_part_size=6b0000
sf_part_start=950000
stderr=serial
stdin=serial
stdout=serial

Environment size: 621/4092 bytes
```

I will try to dump firmware from SPI NOR and will give you an update shortly.

**Photos**:
![1](https://user-images.githubusercontent.com/58137164/171840985-4daa4cae-6bd5-4c34-9bdd-927218dd122c.JPG)
![2](https://user-images.githubusercontent.com/58137164/171840990-e68c9790-0b49-4631-a72f-46812e510289.JPG)
![3](https://user-images.githubusercontent.com/58137164/171840993-09b6354c-335a-4d26-a005-3cd2e05282fc.JPG)
![4](https://user-images.githubusercontent.com/58137164/171840998-7f4affd2-308b-4f94-a996-e611208b68f9.JPG)
![5](https://user-images.githubusercontent.com/58137164/171841003-e1206a28-9f05-4682-a29c-8a753a23c4a8.JPG)
![6](https://user-images.githubusercontent.com/58137164/171841007-5a1408c4-cf11-42a8-94fb-0a46aadd3f6d.JPG)

**Comment by Mat-Alm** [comment]
I think it's better to use an external programmer to dump the nor flash.

**Comment by iamdavidcz** [comment]
Tried to boot Linux 5.10.123, but without a bit of luck. Probably, it has something to do with a difference in uImage structure. Sigmastar vendor kernel injects FDT with `ms_builtin_dtb_update.py` script.
```
$ binwalk -r uImage uImage.ssc325

Scan Time:     2022-06-18 10:55:14
Target File:   sigmastar/uImage
MD5 Checksum:  7982292852736de37b8c6f8c2da80f1c
Signatures:    391

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             uImage header, header size: 64 bytes, header CRC: 0xCA249F8E, created: 2022-06-18 08:14:22, image size: 3297128 bytes, Data Address: 0x20008000, Entry Point: 0x20008000, data CRC: 0xB49BD8C9, OS: Linux, CPU: ARM, image type: OS Kernel Image, compression type: none, image name: "Linux-SSC325"
64            0x40            Linux kernel ARM boot executable zImage (little-endian)
2436          0x984           device tree image (dtb)
14216         0x3788          xz compressed data
14437         0x3865          xz compressed data


Scan Time:     2022-06-18 10:55:14
Target File:   sigmastar/uImage.ssc325
MD5 Checksum:  fa44a6df0147421b04b3df6d4a0a4e7c
Signatures:    391

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             uImage header, header size: 64 bytes, header CRC: 0x71C11797, created: 2022-05-27 03:30:28, image size: 1943992 bytes, Data Address: 0x20008000, Entry Point: 0x20008000, data CRC: 0x9BEEC488, OS: Linux, CPU: ARM, image type: OS Kernel Image, compression type: none, image name: "Linux-4.9.84-ssc325"
2264          0x8D8           device tree image (dtb)
10708         0x29D4          device tree image (dtb)
14016         0x36C0          device tree image (dtb)
14971         0x3A7B          xz compressed data
15305         0x3BC9          xz compressed data
```

**Comment by iamdavidcz** [comment]
@fifteenhex Any chances that your DDR init code in U-Boot will also work in SSC325/335 chips? If yes, could you point me to the right direction how to test it? Could you share your U-Boot tree with DDR init support, please?

**Reply by fifteenhex** [reply]
The code should work but we probably need to adjust it for the SSC325. The issue is that despite the memory probably working with the timing for the SSD210 (both have 64MB DDR2 which is probably exactly the same part or pretty close) they seem to have made detecting the memory type in the chip different for each family..

Basically, we'll need to run my little ghidra emulator thing with the SSC325 ipl and see what it does for DDR setup and put the differences into the u-boot code. The u-boot code is in the u-boot repo if you want to look at the horror. :D
